### PR TITLE
status via streams

### PR DESCRIPTION
**Pull request by elbuo8**
By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/.github/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

> Describe the purpose of this PR along with any background information and the impacts of the proposed change. For the benefit of the community, please do not assume prior context.
>
> Provide details that support your chosen implementation, including: breaking changes, alternatives considered, changes to the API, etc.
>
> If the UI is being changed, please provide screenshots.

### References

> Include any links supporting this change such as a:
>
> - GitHub Issue/PR number addressed or fixed
> - Auth0 Community post
> - StackOverflow post
> - Support forum thread
> - Related pull requests/issues from other repos
>
> If there are no references, simply delete this section.

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors.
>
> Please include any manual steps for testing end-to-end or functionality not covered by unit/integration tests.
>
> Also include details of the environment this PR was developed in (language/platform/browser version).

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not `master`

### CODE REVIEW

File: lib/db.js [hunk 1]
```diff
@@ -314,17 +314,35 @@ class LimitDBRedis extends EventEmitter {
...
+          remaining
+        });
+      });
+    }
+  }
+
+  scanStream(params) {
+    const nodes = this.redis.nodes ? this.redis.nodes('slave') : [this.redis];
+    return mergeStream(nodes.map((node) => {
+      return node.scanStream({
```
Comment (dafortune) [inline lib/db.js]: what happens with the errors of the individual streams? I'm not sure how merge stream works, but in general if you don't listen for them they will end up as uncaughts unless mergestream does

File: lib/db.js [hunk 2]
```diff
@@ -349,44 +367,32 @@ class LimitDBRedis extends EventEmitter {
...
+    stream.on('data', (keys) => {
+      for (let i = 0; i < keys.length; i++) {
+        matches.push(keys[i]);
+        if (matches.length >= count) {
           break;
         }
       }
-      return callback(null, { items, errors });
+      if (matches.length >= count) {
+        stream.end();
```
Comment (dafortune) [inline lib/db.js]: I might be wrong, but I don't think this is fine, the writer side is the one in charge of ending the stream, if the reader side `ends` it the result is undefined I think; you might even end up with a `write after end` error I think.

The reader side should simply stop consuming the stream (`stream.pause()`) that way the writter will eventually stop requesting data and eventually the stream will be GC'd

File: lib/db.js [hunk 3]
```diff
@@ -349,44 +367,32 @@ class LimitDBRedis extends EventEmitter {
...
+      }
+    });
+
+    stream.on('end', () => {
+      async.map(matches, transform, (err, items) => {
```
Comment (dafortune) [inline lib/db.js]: How much parallelism do we have here? Is there any limit or it is full series?

Comment (dafortune) [reply lib/db.js]: Why not to use a stream transform and transform on the fly instead of waiting till end?

Comment (dafortune) [reply lib/db.js]: What happens with the errors while transforming?

File: lib/db.js (hunk 1 above)
Comment (elbuo8) [reply lib/db.js]: Merge stream bubbles them up and listens for them individually.

https://github.com/grncdr/merge-stream/blob/master/index.js#L28

File: lib/db.js (hunk 2 above)
Comment (elbuo8) [reply lib/db.js]: Yep, that error would get bubbled up to the on error listener.

Looking for a way to exit quicker. Pause will simply stop calling the data listener but will actually collect data internally thus wasting CPU, memory and Redis calls (particularly this one :D).

Comment (elbuo8) [reply lib/db.js]: I believe since its a Passthrough stream I could use `destroy`.

File: lib/db.js (hunk 3 above)
Comment (elbuo8) [reply lib/db.js]: Swapped.

File: lib/db.js [hunk 4]
```diff
@@ -349,45 +358,36 @@ class LimitDBRedis extends EventEmitter {
...
+
+    rStream.pipe(flattenStream).pipe(keyInfoStream)
+      .on('data', (data) => {
+        if (items.length < count) {
+          items.push(this.buildStatusResponse(bucket, data));
         }
-        if (items.length === count) {
-          break;
+        if (items.length >= count) {
+          rStream.destroy();
```
Comment (dafortune) [inline lib/db.js]: This is not available in node 6. It was added in node 8. Are you sure it is fine to use it?

Comment (dafortune) [reply lib/db.js]: We might want too reduce the count and add a test case for this if it is not already there.

Comment (elbuo8) [reply lib/db.js]: Solid catch. No bueno.

File: lib/db.js [hunk 5]
```diff
@@ -349,45 +358,36 @@ class LimitDBRedis extends EventEmitter {
...
-      const items = [];
-      for (let i = 0; i < results.length; i++) {
-        for (let j = 0; j < results[i].length; j++) {
-          const match = results[i][j];
-          items.push(transform(match[2], parseInt(match[0], 10), parseInt(match[1], 10)));
-          if (items.length === count) {
-            break;
-          }
+    const items = [];
+    let errors = [];
```
Comment (dafortune) [inline lib/db.js]: This might be a const right?

File: lib/db.js (hunk 4 above)
Comment (elbuo8) [reply lib/db.js]: It's already there the test. But `destroy` not being available in 6 makes it... :(

**Pull request closed without merging**